src/sentry/static/sentry/app/views/projectKeyDetails.jsx

```diff
@@ -2,43 +2,179 @@ import React from 'react';
+
+    let value = `${accepted.toLocaleString()} accepted`;
+    if (dropped) {
+      value += `<br>${dropped.toLocaleString()} rate limited`;
+    }
+    if (filtered) {
+      value += `<br>${filtered.toLocaleString()} filtered`;
+    }
+
+    return (
+      '<div style="width:150px">' +
+      `<div class="time-label">${timeLabel}</div>` +
+      `<div class="value-label">${value}</div>` +
+      '</div>'
+    );
+  },
+
+  render() {
+    if (this.state.loading) return <div className="box"><LoadingIndicator /></div>;
+    else if (this.state.error) return <LoadingError onRetry={this.fetchData} />;
+
+    return (
+      <div className="panel panel-default">
+        <div className="panel-heading">
+          <h6>{t('Key usage in the last 30 days (by day)')}</h6>
+        </div>
+        <div className="panel-body p-a-0">
+          {!this.state.emptyStats
+            ? <div className="inbound-filters-stats p-a-1">
+                <div className="bar-chart">
+                  <StackedBarChart
+                    points={this.state.stats}
+                    height={100}
+                    barClasses={['accepted', 'rate-limited', 'black-listed']}
+                    className="sparkline m-b-0"
+                    tooltip={this.renderTooltip}
+                  />
+                </div>
+              </div>
+            : <div className="blankslate p-y-2">
+                <h5>{t('Nothing recorded in the last 30 days.')}</h5>
+                <p className="m-b-0">
+                  {t('Total events captured using these credentials.')}
+                </p>
+              </div>}
+        </div>
+      </div>
+    );
+  }
+});
+
 const KeySettings = React.createClass({
   propTypes: {
+    organization: React.PropTypes.object.isRequired,
+    project: React.PropTypes.object.isRequired,
     access: React.PropTypes.object.isRequired,
-    orgId: React.PropTypes.string.isRequired,
-    projectId: React.PropTypes.string.isRequired,
     data: React.PropTypes.object.isRequired,
     initialData: React.PropTypes.object,
     onRemove: React.PropTypes.func.isRequired,
-    onSave: React.PropTypes.func.isRequired
+    onSave: React.PropTypes.func.isRequired,
+    rateLimitsEnabled: React.PropTypes.bool
   },
 
   mixins: [ApiMixin],
 
   getInitialState() {
+    let hooksDisabled = [];
+    HookStore.get('project:rate-limits:disabled').forEach(cb => {
+      hooksDisabled.push(cb);
+    });
     return {
       formData: Object.assign({}, this.props.initialData),
-      errors: {}
+      errors: {},
+      hooksDisabled: hooksDisabled
     };
   },
 
   onFieldChange(name, value) {
-    this.setState({
-      formData: {
-        ...this.state.formData,
-        [name]: value
-      }
+    this.setState(state => {
+      return {
+        formData: {
+          ...state.formData,
+          [name]: value
+        }
+      };
+    });
+  },
+
+  onRateLimitChange(name, value) {
+    this.setState(state => {
```

Comment:
Future thinking: we should maybe explore using [immutability-helper](https://github.com/kolodny/immutability-helper) for these types of updates.